src/Compilers/Core/Portable/SourceGeneration/RunResults.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        public ImmutableArray<GeneratorRunResult> Results { get; }
+
+        /// <summary>
+        /// The <see cref="Diagnostic"/>s produced by all generators run during this generation pass.
+        /// </summary>
+        /// <remarks>
+        /// This is equivalent to the union of all <see cref="GeneratorRunResult.Diagnostics"/> in <see cref="Results"/>.
+        /// </remarks>
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        /// <summary>
+        /// The <see cref="SyntaxTree"/>s produced during this generation pass by parsing each <see cref="SourceText"/> added by each generator.
+        /// </summary>
+        /// <remarks>
+        /// This is equivalent to the union of all <see cref="GeneratedSourceResult.SyntaxTree"/>s in each <see cref="GeneratorRunResult.AddedSources"/> in each <see cref="Results"/>
+        /// </remarks>
+        public ImmutableArray<SyntaxTree> SyntaxTrees { get; }
+    }
+
+    /// <summary>
+    /// Represents the results of a single <see cref="ISourceGenerator"/> generation pass.
+    /// </summary>
+    public readonly struct GeneratorRunResult
+    {
+        internal GeneratorRunResult(ISourceGenerator generator, ImmutableArray<GeneratedSourceResult> addedSources, ImmutableArray<Diagnostic> diagnostics, Exception? exception)
+        {
+            this.Generator = generator;
+            this.AddedSources = addedSources;
+            this.Diagnostics = diagnostics;
+            this.Exception = exception;
+        }
+
+        /// <summary>
+        /// The <see cref="ISourceGenerator"/> that produced this result.
+        /// </summary>
+        public ISourceGenerator Generator { get; }
+
+        /// <summary>
+        /// The sources that were added by <see cref="Generator"/> during the generation pass this result represents.
+        /// </summary>
+        public ImmutableArray<GeneratedSourceResult> AddedSources { get; }
+
+        /// <summary>
+        /// A collection of <see cref="Diagnostic"/>s reported by <see cref="Generator"/> 
+        /// </summary>
+        /// <remarks>
+        /// When generation fails due to an <see cref="Exception"/> being thrown, a single diagnostic is added
+        /// to represent the failure. Any generator reported diagnostics up to the failure point are not included.
+        /// </remarks>
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        /// <summary>
+        /// An <see cref="System.Exception"/> instance that was thrown by the generator, or <c>null</c> if the generator completed without error.
+        /// </summary>
+        /// <remarks>
+        /// When this property has a value, <see cref="AddedSources"/> property is guaranteed to be empty, and the <see cref="Diagnostics"/>
+        /// collection will contain a single diagnostic indicating that the generator failed.
+        /// </remarks>
+        public Exception? Exception { get; }
```

Comment:
Want to make this an array of Exceptions should we ever need more than one for some reason? I'm not sure if you'll have cases where you're calling a generator in parallel and might have multiple exceptions being thrown simultaneously, for example.

Reply:
That sounds like a case for everyone's best friend: `AggregateException`